[ModuleInterface] Print full type if ambiguous for extensions.

varungandhi-apple opened the pull request:
Fixes rdar://79093752.

Not sure why `Options.FullyQualifiedTypesIfAmbiguous` was set to `false` in the first place (I would think that setting should always be true apart from debugging?).

varungandhi-apple commented:
@swift-ci smoke test

varungandhi-apple commented:
Okay, this breaks IDE tests because more things start getting qualified... trying to figure out a different solution...

slavapestov commented:
I believe there's a way to set PrintOptions specifically for printing module interfaces which doesn't affect the IDE tests. I agree that the extended type should always be fully qualified in module interfaces, just like we fully qualify TypeReprs in other contexts as well.

varungandhi-apple committed:
[ModuleInterface] Print full type if ambiguous for extensions.

The patch introduces a new setting instead of changing existing settings
because the generated interfaces in the IDE have slightly different
requirements; the extended type there is unconditionally not printed
qualified (even if it is ambiguous). This is likely because the
ambiguity heuristic is very weak; it doesn't even do name lookup.
Simplifying that logic would be nice, but then we'd need to update
a bunch of IDE/print* tests and end up with more more visual clutter
in the IDE.

Introducing the new setting means we can change the behavior for
swiftinterface files without affecting the behavior for IDE interfaces.

Fixes rdar://79093752.

varungandhi-apple commented:
One of the things need to make this work is to remove the forced `= false;` in `printExtendedTypeName` and one more function (so that any changes in `printSwiftInterfaceFile` would propagate correctly). However, the problem with doing that is that the IDE setting for `FullyQualifiedTypesIfAmbiguous` + weak ambiguity detection means that the IDE output still ended up changing (presumably, that's why the `= false` was present?). So I've added a new setting specifically for extended types to preserve the IDE output.

If we added better ambiguity detection in the future, then we could remove this extra setting and still have nice IDE output in the common case.

varungandhi-apple commented:
@swift-ci smoke test

varungandhi-apple commented on test/SourceKit/CodeComplete/complete_swiftinterface.swift:
```diff
@@ -19,10 +19,11 @@ let x = MyPoint(x: 1, y: 10.5)
 // CHECK: Decl[InstanceVar]/CurrNominal:      magnitudeSquared[#Double#]; name=magnitudeSquared
-// CHECK: Decl[InstanceVar]/CurrNominal:      magnitude[#Double#]; name=magnitude
+// FIXME: The following completion is missing.
+// CHECK-NOT: Decl[InstanceVar]/CurrNominal:      magnitude[#Double#]; name=magnitude
```
False alarm, it's not a regression, it was the "classic" problem of module name = type name, which confused SourceKit.

varungandhi-apple commented:
@swift-ci smoke test

varungandhi-apple commented:
@swift-ci smoke test

https://github.com/apple/llvm-project/pull/3040

varungandhi-apple commented:
@swift-ci test

https://github.com/apple/llvm-project/pull/3040

varungandhi-apple commented:
@swift-ci clean test

apple/llvm-project#3040

varungandhi-apple commented:
@swift-ci clean test

varungandhi-apple commented:
@swift-ci test

apple/llvm-project#3040

varungandhi-apple commented:
apple/llvm-project#3040

@swift-ci test

varungandhi-apple commented:
apple/llvm-project#3040

@swift-ci test

varungandhi-apple commented:
apple/llvm-project#3040

@swift-ci test macOS

varungandhi-apple commented:
Got LGTM separately from Slava.
